### PR TITLE
append code coverage over each other

### PR DESCRIPTION
**Pull request by seankane-msft**

### CODE REVIEW

File: eng/pipelines/templates/stages/archetype-sdk-tests.yml [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        AllocateResourceGroup: ${{ parameters.AllocateResourceGroup }}
+        DeployArmTemplate: ${{ parameters.DeployArmTemplate }}
+        TestTimeoutInMinutes: ${{ parameters.TestTimeoutInMinutes }}
+        TestSamples: ${{ parameters.TestSamples }}
+        Location: ${{ parameters.Location }}
+        Matrix: ${{ parameters.Matrix }}
+        CloudConfigurations: ${{ parameters.CloudConfigurations }}
+
+
+  - stage: Publish_Coverage
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-tests.yml:110]: Why make this a separate stage/job?

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml [hunk 2]
```diff
@@ -1,49 +1,85 @@
 parameters:
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml:1]: FYI @benbp I know you are working on refactoring this file and so you will likely conflict with this PR.

Comment (benbp) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:1]: @weshaggard I chatted with Sean and was waiting for these changes to land first, hence why I did Java instead.

File: eng/pipelines/templates/stages/archetype-sdk-tests.yml [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  - name: TestSamples
+    type: boolean
+    default: false
+  - name: Location
+    type: string
+    default: ''
+  - name: Matrix
+    type: object
+    default:
+      Linux_Python35:
```
Comment (benbp) [inline eng/pipelines/templates/stages/archetype-sdk-tests.yml:53]: This is duplicated with the job parameters. I think you can keep it here and remove it there (though I'm going to delete both subsequently, so its optional)

File: eng/pipelines/templates/stages/archetype-sdk-tests.yml (hunk 1 above)
Comment (benbp) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:110]: I would call the stage something other than `Build`. Maybe `LiveTest` or just leave it blank

File: eng/pipelines/templates/stages/archetype-sdk-tests.yml [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        DeployArmTemplate: ${{ parameters.DeployArmTemplate }}
+        TestTimeoutInMinutes: ${{ parameters.TestTimeoutInMinutes }}
+        TestSamples: ${{ parameters.TestSamples }}
+        Location: ${{ parameters.Location }}
+        Matrix: ${{ parameters.Matrix }}
+        CloudConfigurations: ${{ parameters.CloudConfigurations }}
+
+
+  - stage: Publish_Coverage
+    dependsOn: Build
```
Comment (benbp) [inline eng/pipelines/templates/stages/archetype-sdk-tests.yml:111]: The dependsOn is unnecessary, since it will implicitly do the stages in list order. But as Wes asked, you can probably just have this in the job list instead.

File: eng/pipelines/templates/stages/archetype-sdk-tests.yml (hunk 1 above)
Comment (seankane-msft) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:110]: @weshaggard the code coverage is dependent on all the Python 3.9 jobs.

File: eng/pipelines/templates/jobs/publish-coverage.yml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+      - task: PublishCodeCoverageResults@1
+        displayName: 'Publish Code Coverage to DevOps'
+        continueOnError: true
+        condition: and(succeededOrFailed(), ${{ parameters.RunCoverage }})
+        inputs:
+          codeCoverageTool: Cobertura
+          summaryFileLocation: '$(Build.SourcesDirectory)/coverage.xml'
+
+  - job: Show_all_coverage_files
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/publish-coverage.yml:65]: Why not do this as another step in the previous job? We don't want to use another agent machine to just do that.

File: eng/pipelines/templates/stages/archetype-sdk-tests.yml (hunk 1 above)
Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:110]: So this this is an aggregate coverage report publishing? Is there advantage to do it this way as opposed to letting each job publish the results individually? If there is an advantage to doing an aggregate then this approach should work but I'd suggest updating the name to indicate that is what is doing.

Comment (seankane-msft) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:110]: If each job publishes individually they will overwrite previous reports. This method guarantees that all jobs have an accurate code coverage report.

Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:110]: I'm not opposed to doing an aggregate report and publishing but just curious why we couldn't solve this problem by giving each job a unique name?

Comment (seankane-msft) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:110]: @weshaggard yes, the code coverage reports are currently over-writing each other each time they are published. This ensures that one report is published with all the code coverage for the package in one place.

File: eng/pipelines/templates/steps/build-test.yml [hunk 6]
```diff
@@ -55,17 +55,29 @@ steps:
...
+      }
+    displayName: 'Show .coverage files'
+    condition: and(succeededOrFailed(), ${{ parameters.RunCoverage }})
+
+  - task: PublishPipelineArtifact@1
+    displayName: 'Publish .coverage files'
+    continueOnError: true
+    condition: and(succeededOrFailed(), ${{ parameters.RunCoverage }})
+    inputs:
+      targetPath: '$(Build.SourcesDirectory)/_coverage'
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/build-test.yml:71]: Does this work? DevOps generally doesn't like you uploading an artifact with the same name from multiple jobs/stages. Usually you have to have a unique name.

Comment (seankane-msft) [reply eng/pipelines/templates/steps/build-test.yml:71]: Yes it uploads the artifacts in a folder named after the job so each job is unique.

File: eng/pipelines/templates/jobs/publish-coverage.yml [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        continueOnError: false
+        condition: and(succeededOrFailed(), ${{ parameters.RunCoverage }})
+
+      - task: DownloadPipelineArtifact@2
+        inputs:
+          source: current
+          path: '$(Build.SourcesDirectory)/_all_coverage_files'
+
+      - script: |
+          python -m pip install pip==20.1
```
Comment (scbedd) [inline eng/pipelines/templates/jobs/publish-coverage.yml:40]: Can we instead place `pip==20.2` in the `eng/ci_tools.txt`? That way it's pinned in an expected location.

File: sdk/storage/tests_invoke.yml [hunk 8]
```diff
@@ -3,7 +3,7 @@ parameters:
   InjectedPackages: ''
 
 jobs:
-  - template: ../../eng/pipelines/templates/jobs/archetype-sdk-tests.yml
```
Comment (scbedd) [inline sdk/storage/tests_invoke.yml:6]: Is this file necessary anymore? We removed the reference to tests_invoke.yml in the storage tests.yml IIRC.

Comment (scbedd) [reply sdk/storage/tests_invoke.yml:6]: Aha. I'd prefer that `tests.yml` call into `tests_invoke.yml`. 

The reason that it was set up a bit non-standard was to minimize the duplication of the environment variables being passed to `archetype-sdk-tests`.

File: scripts/devops_tasks/tox_harness.py [hunk 9]
```diff
@@ -112,7 +112,7 @@ def collect_tox_coverage_files(targeted_packages):
 
     # coverage report has paths starting .tox and azure
     # coverage combine fixes this with the help of tox.ini[coverage:paths]
-    combine_coverage_files(targeted_packages)
+    # combine_coverage_files(targeted_packages) # Skipping this, upload all .coverage-{package-name} files
```
Comment (scbedd) [inline scripts/devops_tasks/tox_harness.py:115]: I think we need agree on how `coverage` artifacts will be handled. Specifically, in some places, we comment out `combine` of coverage. However, we still talk about as if we were combining down below

```
logging.info("Visible uncombined .coverage files: {}".format(coverage_files))
```

**Review by scbedd — CHANGES_REQUESTED** [review]
Couple structural questions to do with storage live tests. Need to clean up the bounds around `coverage` in setup_execute and tox_harness. 

The vast majority is great! Couple questions and let's get this merged.

File: scripts/devops_tasks/create_coverage.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        run_check_call(commands, root_dir, always_exit = False)
+    else:
+        logging.error("Coverage file is not available in {} to generate coverage XML".format(coverage_dir))
+
+
+def fix_dot_coverage_file(coverage_file):
+    print("running 'fix_dot_coverage_file' on {}".format(coverage_file))
+
+    out = None
+    with open(coverage_file) as cov_file:
```
Comment (scbedd) [inline scripts/devops_tasks/create_coverage.py:63]: This specific function makes me side-eye it.

I know you've been dumping content in the publish-coverage job. You know what's up, but this specific line really scares me.

Can you talk me through it in a quick call?

File: scripts/devops_tasks/tox_harness.py (hunk 9 above)
Comment (seankane-msft) [reply scripts/devops_tasks/tox_harness.py:115]: Changed this to logging information about what files are found.

File: sdk/storage/tests_invoke.yml (hunk 8 above)
Comment (seankane-msft) [reply sdk/storage/tests_invoke.yml:6]: fixed!

File: scripts/devops_tasks/create_coverage.py (hunk 10 above)
Comment (seankane-msft) [reply scripts/devops_tasks/create_coverage.py:63]: Result: change the replace to replace anything with the regex `/.tox/**/site-packages` to a blank string

**Review by scbedd — APPROVED** [review]
Pending our discussed items. G2g.

File: eng/pipelines/templates/jobs/publish-coverage.yml [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+  - name: ServiceDirectory
+    type: string
+    default: ''
+  - name: Artifacts
+    type: object
+    default: []
+  - name: DependsOn
+    type: string
+    default: ''
+  - name: RunCoverage
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/publish-coverage.yml:11]: Is this interesting any more? This step isn't doing any running but only publishing.

Comment (seankane-msft) [reply eng/pipelines/templates/jobs/publish-coverage.yml]: Nope, removing

File: eng/pipelines/templates/jobs/publish-coverage.yml [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        displayName: Show all coverage files
+
+      - task: PythonScript@0
+        displayName: 'Create Coverage Report'
+        inputs:
+          scriptPath: 'scripts/devops_tasks/create_coverage.py'
+
+      - pwsh: |
+          Get-Content $(Build.SourcesDirectory)\.coverage
+        continueOnError: true
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/publish-coverage.yml:57]: If you If you plan to keep these steps you should given them a displayname so it is clear what they are doing in the logs and it might be worth combining these 2.

File: sdk/cosmos/tests.yml [hunk 13]
```diff
@@ -1,7 +1,7 @@
 trigger: none
 
 jobs:
```
Comment (benbp) [inline sdk/cosmos/tests.yml:3]: You need this top level `jobs` field to be renamed to `stages` too for all these `tests.yml` files. Looks like you did it for a few services but not for the majority.

**Review by benbp — DISMISSED** [review]
